linkerd/stack/src/lib.rs

```diff
@@ -38,3 +38,20 @@ pub use self::{
 };
 pub use tower::util::Either;
 pub use tower::util::{future_service, FutureService};
+
+/// Describes a stack target that can produce `T` typed parameters.
+///
+/// Stacks (usually layered `NewService` implementations) frequently need to be
+/// able to obtain configuration from the stack target, but stack modules are
+/// decoupled from any concrate target types. The `Param` trait
+pub trait Param<T> {
+    /// Produces a `T`-typed stack paramter.
+    fn param(&self) -> T;
+}
```

Comment:
Do we care about being able to have `Param` impls where `T` is borrowed from `Self`? In order to support that we'd need to change this to something like
```rust
pub trait Param<'a, T> {
   fn param(&'a self) -> T;
}
```
so you can write impls like 
```rust
pub struct Foo {
    bar: Bar,
}

impl<'a> Param<&'a Bar> for Foo {
    fn param(&'a self) -> &'a Bar {
       &self.bar
    }
}
```

On the other hand that complicates both the implementation and the usage --- the `for<'a>` probably has to come back, and getting rid of that was part of the motivation for this change...

Reply:
I don't think it's worth the complication. If we find a need for it, we can introduce a `ParamRef` trait; but, generally, we accept that target types are cloned freely, so it's up to the types to make that cheap (i.e. by using Arc where appropriate).